[TextureMapper] Polygon clipping performance is extremely slow

zhani opened the pull request:
#### 3440db247780d0361df3c48b07059f57ca86e92b
<pre>
[TextureMapper] Polygon clipping performance is extremely slow
<a href="https://bugs.webkit.org/show_bug.cgi?id=284027">https://bugs.webkit.org/show_bug.cgi?id=284027</a>

Reviewed by Fujii Hironori.

Rendering complex preserve-3D scenes might involve a lot of layer splitting.
These splits are handled using polygon clipping, but the current implementation
is inefficient. Each clip operation clears the full stencil buffer and masks
the clipped part using separate vertex buffer objects that are allocated on
demand. This places significant stress on the GPU and on hardware like rpi3 it
completely chokes the rendering pipeline.

To address this issue, polygon clip vertices for preserve-3d scenes should be
loaded to the GPU only once, leveraging cached VBOs. Additionally, stencil
buffer clearing should be optimized using scissors to restrict stencil clears to
relevant areas, avoiding full-buffer operations.

* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/texmap/ClipPath.cpp: Added.
(WebCore::ClipPath::ClipPath):
(WebCore::ClipPath::data const):
* Source/WebCore/platform/graphics/texmap/ClipPath.h: Added.
* Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
(WebCore::nextPowerOf2):
(WebCore::TextureMapperGLData::initializeStencil):
(WebCore::TextureMapperGLData::getBufferFromPool):
(WebCore::TextureMapper::beginClip):
(WebCore::TextureMapper::updateProjectionMatrix):
(WebCore::TextureMapper::acquireBufferFromPool):
* Source/WebCore/platform/graphics/texmap/TextureMapper.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperGPUBuffer.cpp: Added.
(WebCore::TextureMapperGPUBuffer::TextureMapperGPUBuffer):
(WebCore::TextureMapperGPUBuffer::~TextureMapperGPUBuffer):
(WebCore::TextureMapperGPUBuffer::updateData):
* Source/WebCore/platform/graphics/texmap/TextureMapperGPUBuffer.h: Added.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::paintWith3DRenderingContext):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
(WebCore::TextureMapperLayer3DRenderingContext::paint):
(WebCore::TextureMapperLayer3DRenderingContext::traverseTree):
(WebCore::TextureMapperLayer3DRenderingContext::traverseTreeAndPaint): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:

Canonical link: <a href="https://commits.webkit.org/287519@main">https://commits.webkit.org/287519@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8638a340e7c8d4a4077e4faf2e7156fa659131de

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79979 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58976 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33377 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84493 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30953 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82087 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68040 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7272 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62514 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20340 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83046 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/52577 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72839 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42823 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49915 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26992 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29415 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71035 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27483 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85925 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7195 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5062 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70786 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7372 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68680 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70030 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14026 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12961 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12368 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7159 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12693 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7006 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10518 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8810 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebCore/platform/graphics/texmap/ClipPath.cpp:
```diff
@@ -0,0 +1,60 @@
+#include "ClipPath.h"
+
+#include "FloatRect.h"
+
```
WebKit doesn't put a blank line here.

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/ClipPath.h:
```diff
@@ -0,0 +1,59 @@
+class ClipPath final {
+    WTF_MAKE_TZONE_ALLOCATED(ClipPath);
+public:
+    ClipPath(Vector<FloatPoint>&&, unsigned, unsigned);
```
Needs parameter names.
> ClipPath(Vector<FloatPoint>&& vertices, unsigned bufferID, unsigned bufferOffsetInBytes);

https://webkit.org/code-style-guidelines/#names-variable-name-in-function-decl

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/ClipPath.h:
```diff
@@ -0,0 +1,59 @@
+
+#include "FloatPoint.h"
+#include "FloatRect.h"
+
```
WebKit doesn't put a blank line here.

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
```diff
@@ -169,6 +188,28 @@ Ref<TextureMapperShaderProgram> TextureMapperGLData::getShaderProgram(TextureMap
+        return zeroBuffer;
+    }
+
+    uint64_t key = (static_cast<uint64_t>(type) << 32) | static_cast<uint32_t>(size);
```
You truncate 64bit to 32bit silently here. If size is very big, this function can return a small size buffer.
Let's add RELEASE_ASSERT(size < std::numeric_limits<uint32_t>::max());

zhani replied:
Added. It's extremely unlikely to happen in practice, even though possible in theory.

To make it fail safe in both 32bit and 64bit cases, `size` could be splitted into two 32bit parts when `sizeof(size_t) > 4` and refactor key calculation a bit

fujii reviewed (changes requested):
I will check more later.

fujii commented on Source/WebCore/platform/graphics/texmap/ClipPath.cpp:
```diff
@@ -0,0 +1,58 @@
+    if (m_bufferID)
+        return reinterpret_cast<const void*>(m_bufferOffsetInBytes);
+
+    return m_vertices.data();
```
Why do you return m_vertices.data() here? When is m_bufferID is 0?

zhani replied:
At some point I had a plan that ClipPath could be used as a CPU buffer for clip vertices and then uploaded to GPU on demand. I did not proceed with that approach and this is left over from that. Removed now and function name updated to `bufferDataOffsetAsPtr()`

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
```diff
@@ -1329,14 +1371,50 @@ void TextureMapper::beginClip(const TransformationMatrix& modelViewMatrix, const
+    auto clipP1NDC = posToNDC(clipBounds.minXMinYCorner(), mvpMatrix);
+    auto clipP2NDC = posToNDC(clipBounds.maxXMinYCorner(), mvpMatrix);
+    auto clipP3NDC = posToNDC(clipBounds.maxXMaxYCorner(), mvpMatrix);
+    auto clipP4NDC = posToNDC(clipBounds.minXMaxYCorner(), mvpMatrix);
```
Why do you need to convert to NDC temporalily? You convert the NDC points to screen coord. Why not convert them to screen coord and take min/max?

zhani replied:
Initially, I had some issues and random artifacts which led me to take a step-by-step approach to ensure clarity and functionality. I used that code because it is working and had clear steps. However,  now I tried again with `mapQuad` and see no more artifacts. While there are occasional one-pixel differences in `scissorRect` calculations compared to NDC approach, they don’t appear to cause any noticeable issues now.

zhani replied:
I’ve updated the implementation to use the `mapQuad`

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperGPUBuffer.h:
```diff
@@ -0,0 +1,69 @@
+
+namespace WebCore {
+
+class TextureMapperGPUBuffer final : public ThreadSafeRefCounted<TextureMapperGPUBuffer> {
```
I think we should mark this class with WTF_MAKE_NONCOPYABLE.

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
```diff
@@ -1427,6 +1509,16 @@ void TextureMapper::drawTextureExternalOES(GLuint texture, OptionSet<TextureMapp
+    size_t ceil = nextPowerOf2(size);
+    size_t floor = ceil >> 1; // half of ceil
+    size_t mid = floor + (floor >> 1); // (1.5 times floor)
+    size_t requestSize = (size <= mid) ? mid : ceil;
```
You can use `std::countl_zero` to simplify this code.
https://godbolt.org/z/jY8xvP6eG

zhani replied:
`std::countl_zero` is new to me. I applied it to `nextPowerOf2`, but I'm not entirely sure if that's the specific usage you had in mind.

zhani committed:
[TextureMapper] Polygon clipping performance is extremely slow
https://bugs.webkit.org/show_bug.cgi?id=284027

Reviewed by Fujii Hironori.

Rendering complex preserve-3D scenes might involve a lot of layer splitting.
These splits are handled using polygon clipping, but the current implementation
is inefficient. Each clip operation clears the full stencil buffer and masks
the clipped part using separate vertex buffer objects that are allocated on
demand. This places significant stress on the GPU and on hardware like rpi3 it
completely chokes the rendering pipeline.

To address this issue, polygon clip vertices for preserve-3d scenes should be
loaded to the GPU only once, leveraging cached VBOs. Additionally, stencil
buffer clearing should be optimized using scissors to restrict stencil clears to
relevant areas, avoiding full-buffer operations.

* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/texmap/ClipPath.cpp: Added.
(WebCore::ClipPath::ClipPath):
(WebCore::ClipPath::data const):
* Source/WebCore/platform/graphics/texmap/ClipPath.h: Added.
* Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
(WebCore::nextPowerOf2):
(WebCore::TextureMapperGLData::initializeStencil):
(WebCore::TextureMapperGLData::getBufferFromPool):
(WebCore::TextureMapper::beginClip):
(WebCore::TextureMapper::updateProjectionMatrix):
(WebCore::TextureMapper::acquireBufferFromPool):
* Source/WebCore/platform/graphics/texmap/TextureMapper.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperGPUBuffer.cpp: Added.
(WebCore::TextureMapperGPUBuffer::TextureMapperGPUBuffer):
(WebCore::TextureMapperGPUBuffer::~TextureMapperGPUBuffer):
(WebCore::TextureMapperGPUBuffer::updateData):
* Source/WebCore/platform/graphics/texmap/TextureMapperGPUBuffer.h: Added.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::paintWith3DRenderingContext):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp:
(WebCore::TextureMapperLayer3DRenderingContext::paint):
(WebCore::TextureMapperLayer3DRenderingContext::traverseTree):
(WebCore::TextureMapperLayer3DRenderingContext::traverseTreeAndPaint): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:

Canonical link: https://commits.webkit.org/287519@main
